artichoke-backend/src/extn/core/time/offset.rs

```diff
@@ -0,0 +1,211 @@
+use crate::extn::prelude::*;
+
+#[allow(clippy::cast_precision_loss)]
+const MAX_FLOAT_OFFSET: f64 = i32::MAX as f64;
+#[allow(clippy::cast_precision_loss)]
+const MIN_FLOAT_OFFSET: f64 = i32::MIN as f64;
+
+impl TryConvertMut<Value, Option<Offset>> for Artichoke {
+    type Error = Error;
+
+    fn try_convert_mut(&mut self, options: Value) -> Result<Option<Offset>, Self::Error> {
+        let hash: Vec<(Value, Value)> = self.try_convert_mut(options)?;
+
+        // Short circuit. An empty options hash does not error.
+        //
+        // Example:
+        //
+        // ```console
+        // [2.6.3]> Time.at(0, {})
+        // => 1970-01-01 01:00:00 +0100
+        // ```
+        if hash.is_empty() {
+            return Ok(None);
+        }
+
+        // All other keys are rejected.
+        //
+        // Example:
+        // ```console
+        // [2.6.3]> Time.at(0, i: 0)
+        // ArgumentError (unknown keyword: i)
+        // ```
+        for (mut key, _) in &hash {
+            let k = unsafe { Symbol::unbox_from_value(&mut key, self)? }.bytes(self);
+            if k != b"in" {
+                let mut message = b"unknown keyword: ".to_vec();
+                message.extend_from_slice(k);
+                Err(ArgumentError::from(message))?;
+            }
+        }
+
+        // Based on the above logic, the only option in the hash is `in`.
+        // >0 keys, and all other keys are rejected).
+        let mut in_value = hash.get(0).expect("Only the `in` parameter should be available").1;
+
+        match in_value.ruby_type() {
+            Ruby::String => {
+                let offset_str = unsafe { implicitly_convert_to_string(self, &mut in_value) }?;
+
+                let offset = Offset::try_from(offset_str).map_err(|_| {
+                    let mut message =
+                        b"\"+HH:MM\", \"-HH:MM\", \"UTC\" or \"A\"..\"I\",\"K\"..\"Z\" expected for utc_offset: "
+                            .to_vec();
+                    message.extend_from_slice(offset_str);
+                    ArgumentError::from(message)
+                })?;
+
+                Ok(Some(offset))
+            }
+            Ruby::Float => {
+                // This impl differs from MRI. MRI supports any float value and
+                // will set an offset with subsec fractions however this is not
+                // supported in spinoso_time with the `tzrs` feature.
+                let offset_seconds: f64 = self.try_convert(in_value)?;
+
+                if (MIN_FLOAT_OFFSET..=MAX_FLOAT_OFFSET).contains(&offset_seconds) {
+                    #[allow(clippy::cast_possible_truncation)]
+                    Ok(Some(Offset::try_from(offset_seconds as i32)?))
+                } else {
+                    Err(ArgumentError::with_message("utc_offset out of range").into())
+                }
+            }
+            _ => {
+                let offset_seconds = implicitly_convert_to_int(self, in_value).and_then(|seconds| {
```

Comment:
we need to short circuit return from `implicitly_convert_to_int` here to handle this case:

```
[3.1.2] > class A; def to_int; raise 'bam'; end; end
=> :to_int
[3.1.2] > Time.at(12, in: A.new)
(irb):6:in `to_int': bam (RuntimeError)
        from <internal:timev>:274:in `at'
        from (irb):7:in `<main>'
        from /usr/local/var/rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/irb-1.4.1/exe/irb:11:in `<top (required)>'
        from /usr/local/var/rbenv/versions/3.1.2/bin/irb:25:in `load'
        from /usr/local/var/rbenv/versions/3.1.2/bin/irb:25:in `<main>'
```

Reply:
Filed:

- https://github.com/artichoke/artichoke/issues/2094